Source/DafnyLanguageServer/Workspace/ChangeProcessors/Relocator.cs

```diff
@@ -58,6 +73,22 @@ CancellationToken cancellationToken
         return originalDiagnostics.SelectMany(diagnostic =>
           MigrateDiagnostic(change.Range, change.Text, diagnostic)).ToList();
       }
+      private List<Position> MigratePositions(List<Position> originalRanges, TextDocumentContentChangeEvent change) {
+        if (change.Range == null) {
+          return new List<Position> { };
+        }
+
+        return originalRanges.SelectMany(position => MigratePosition(change.Range, change.Text, position)).ToList();
+      }
+
+      private IEnumerable<Position> MigratePosition(Range changeRange, string changeText, Position position) {
+        var afterChangeEndOffset = GetPositionAtEndOfAppliedChange(changeRange, changeText);
```

Comment:
> GetPositionAtEndOfAppliedChange is potentially expensive, since it scales with the size of the change and a customer could paste in a large file, so I think we should ensure it only gets called once per change. MigratePositions calls it for each position. Maybe we can compute GetPositionAtEndOfAppliedChange in the constructor of ChangeProcessor ?

Done.